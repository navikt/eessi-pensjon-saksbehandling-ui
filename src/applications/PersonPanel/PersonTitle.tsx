import kvinne from 'src/assets/icons/icon-kvinne.png'
import mann from 'src/assets/icons/icon-mann.png'
import ukjent from 'src/assets/icons/icon-ukjent.png'
import classNames from 'classnames'
import { PersonPDL } from 'src/declarations/person'
import _ from 'lodash'
import moment from 'moment'
import {Heading, HStack, Link} from '@navikt/ds-react'
import {getFnr, getNPID} from 'src/applications/BUC/components/BUCUtils/BUCUtils'
import LoadingImage from "src/components/Loading/LoadingImage";
import {copyToClipboard} from "src/actions/app";
import {useDispatch} from "react-redux";
import {useState} from "react";
import {useTranslation} from "react-i18next";
import {FilesFillIcon, FilesIcon} from "@navikt/aksel-icons";
import styles from "src/assets/css/common.module.css";
import titleStyles from "./PersonTitle.module.css"

export interface PersonTitleProps {
  gettingPersonInfo: boolean
  person?: PersonPDL
}

const PersonTitle: React.FC<PersonTitleProps> = ({
  gettingPersonInfo, person
}: PersonTitleProps): JSX.Element => {
  const [hover, isHover] = useState<boolean>(false);
  const dispatch = useDispatch()
  const { t } = useTranslation()

  let birthDate: Date | undefined
  let deathDate: Date | undefined

  let age: number = 0

  if (!person || gettingPersonInfo) {
    return (
      <LoadingImage />
    )
  }

  if (_.get(person, 'foedselsdato.foedselsdato')) {
    birthDate = moment(person.foedselsdato.foedselsdato)?.toDate()
  }
  if (_.get(person, 'doedsfall.doedsdato')) {
    deathDate = moment(person.doedsfall?.doedsdato)?.toDate()
  }

  const endDate: Date = deathDate || new Date()
  if (birthDate && endDate) {
    age = moment(endDate).diff(moment(birthDate), 'years')
  }

  let kind: string = 'nav-unknown-icon'
  let src = ukjent

  if (person.kjoenn?.kjoenn === 'KVINNE') {
    kind = 'nav-woman-icon'
    src = kvinne
  } else if (person.kjoenn?.kjoenn === 'MANN') {
    kind = 'nav-man-icon'
    src = mann
  }

  const pid : string | undefined = getFnr(person) ? getFnr(person) : getNPID(person)

  return (
    <HStack gap="4" align="center">
      <img
        alt={kind}
        className={classNames({ [titleStyles.dead]: !_.isNil(deathDate) })}
        data-testid='w-persontitle--img'
        src={src}
        width={40}
      />
      <Heading size='medium'>
        {person.navn?.sammensattNavn} ({age}) - {pid}
        <Link
          onMouseEnter={() => isHover(true)}
          onMouseLeave={() => isHover(false)}
          title={t('ui:person-kopier-pid')} onClick={(e: any) => {
          e.preventDefault()
          e.stopPropagation()
          pid && dispatch(copyToClipboard(pid))
        }}
        >
          {hover ?
            <FilesFillIcon className={styles.copyFilledWithMargin} fontSize="1.5rem"/> :
            <FilesIcon className={styles.copyWithMargin} fontSize="1.5rem"/>
          }
        </Link>
      </Heading>
    </HStack>
  )
}

export default PersonTitle
