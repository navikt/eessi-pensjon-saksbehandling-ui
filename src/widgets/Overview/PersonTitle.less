.w-overview-personPanel__icon.dead {
  filter: grayscale(100%)
}
