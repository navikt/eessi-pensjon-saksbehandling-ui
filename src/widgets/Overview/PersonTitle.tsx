import kvinne from 'assets/icons/icon-kvinne.png'
import mann from 'assets/icons/icon-mann.png'
import ukjent from 'assets/icons/icon-ukjent.png'
import classNames from 'classnames'
import { HorizontalSeparatorDiv } from 'components/StyledComponents'
import { Person } from 'declarations/types'
import { PersonPropType } from 'declarations/types.pt'
import _ from 'lodash'
import moment, { Moment } from 'moment'
import { Systemtittel } from 'nav-frontend-typografi'
import PT from 'prop-types'
import React from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'
import PersonLoading from 'widgets/Overview/PersonLoading'

export interface PersonTitleProps {
  gettingPersonInfo: boolean
  person?: Person
  personAvdod: any
}

const Title = styled.div`
  display: flex;
  align-items: center;
  .dead {
    filter: grayscale(100%)
  }
`
const PersonTitle: React.FC<PersonTitleProps> = ({
  gettingPersonInfo, person, personAvdod
}: PersonTitleProps): JSX.Element => {
  let birthDate: Date | Moment | undefined
  let deathDate: Date | Moment | undefined
  const { t } = useTranslation()
  if (!person || gettingPersonInfo) {
    return (
      <PersonLoading />
    )
  }

  if (_.get(person, 'foedselsdato.foedselsdato')) {
    birthDate = moment(person.foedselsdato.foedselsdato)
    if (birthDate) {
      birthDate = birthDate.toDate()
    }
  }
  if (_.get(person, 'doedsdato.doedsdato')) {
    deathDate = moment(person.doedsdato.doedsdato)
    if (deathDate) {
      deathDate = deathDate.toDate()
    }
  }

  const age: number = (deathDate ? (deathDate as Date).getFullYear() : new Date().getFullYear()) - (birthDate as Date).getFullYear()

  let kind: string = 'nav-unknown-icon'
  let src = ukjent
  if (person.kjoenn.kjoenn.value === 'K') {
    kind = 'nav-woman-icon'
    src = kvinne
  } else if (person.kjoenn.kjoenn.value === 'M') {
    kind = 'nav-man-icon'
    src = mann
  }

  return (
    <Title>
      <img
        alt={kind}
        width={40}
        src={src}
        className={classNames({ dead: !_.isNil(deathDate) })}
      />
      <HorizontalSeparatorDiv />
      <Systemtittel>
        {person.personnavn.sammensattNavn} ({age}) - {person.aktoer.ident.ident}
        {personAvdod !== undefined ? ' - ' + t('ui:deceased') + ': ' + personAvdod.length : ''}
      </Systemtittel>
    </Title>
  )
}

PersonTitle.propTypes = {
  gettingPersonInfo: PT.bool.isRequired,
  person: PersonPropType
}

export default PersonTitle
