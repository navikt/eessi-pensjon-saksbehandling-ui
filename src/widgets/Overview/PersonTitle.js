import React from 'react'
import PT from 'prop-types'
import _ from 'lodash'
import classNames from 'classnames'
import { Icons, Nav, WaitingPanel } from 'eessi-pensjon-ui'
import './PersonTitle.css'

const PersonTitle = ({ gettingPersonInfo, person, t }) => {
  let birthDate
  let deathDate

  if (!person) {
    return null
  }

  if (gettingPersonInfo) {
    return <WaitingPanel className='w-overview-personPanel__waiting' message={t('ui:loading')} />
  }

  if (person && person.foedselsdato && person.foedselsdato.foedselsdato) {
    birthDate = new Date(Date.parse(person.foedselsdato.foedselsdato))
  }
  if (person && person.doedsdato && person.doedsdato.doedsdato) {
    deathDate = new Date(Date.parse(person.doedsdato.doedsdato))
  }

  const age = (deathDate ? deathDate.getFullYear() : new Date().getFullYear()) - birthDate.getFullYear()

  let kind = 'nav-unknown-icon'
  if (person.kjoenn.kjoenn.value === 'K') {
    kind = 'nav-woman-icon'
  } else if (person.kjoenn.kjoenn.value === 'M') {
    kind = 'nav-man-icon'
  }

  return (
    <div className='w-overview-personPanel__title'>
      <Icons size={40} kind={kind} className={classNames('w-overview-personPanel__icon', { dead: !_.isNil(deathDate) })} />
      <Nav.Systemtittel className='ml-2'>
        {person.personnavn.sammensattNavn} ({age}) - {person.aktoer.ident.ident}
      </Nav.Systemtittel>
    </div>
  )
}

PersonTitle.propTypes = {
  aktoerId: PT.string,
  gettingPersonInfo: PT.bool,
  person: PT.object,
  t: PT.func.isRequired
}

export default PersonTitle
