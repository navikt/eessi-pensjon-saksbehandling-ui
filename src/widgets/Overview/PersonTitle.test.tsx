import React from 'react'
import PersonTitle, { PersonTitleProps } from './PersonTitle'
import { mount, ReactWrapper } from 'enzyme'
import mockPerson from 'mocks/app/person'

describe('widgets/Overview/PersonTitle', () => {
  let wrapper: ReactWrapper
  const initialMockProps: PersonTitleProps = {
    gettingPersonInfo: false,
    person: mockPerson.person
  }

  beforeEach(() => {
    wrapper = mount(<PersonTitle {...initialMockProps} />)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('Renders', () => {
    expect(wrapper.isEmptyRender()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('Renders waiting spinner if no person is given', () => {
    wrapper = mount(<PersonTitle {...initialMockProps} person={undefined} />)
    expect(wrapper.exists('WaitingPanel')).toBeTruthy()
  })

  it('Has proper HTML structure', () => {
    expect(wrapper.exists('.w-overview-personPanel__title')).toBeTruthy()
    expect(wrapper.find('.w-overview-personPanel__title img').props().kind).toEqual('nav-woman-icon')
    expect(wrapper.find('.w-overview-personPanel__title h2').render().text()).toEqual('HØYSÆTHER NAZAKMIR-MASK (89) - 27072942618')
  })

  it('Shows waiting panel when fetching person', () => {
    wrapper = mount(<PersonTitle {...initialMockProps} gettingPersonInfo />)
    expect(wrapper.exists('.w-overview-personPanel__waiting')).toBeTruthy()
    expect(wrapper.find('.w-overview-personPanel__waiting').hostNodes().render().text()).toEqual('Venter...ui:loading')
  })

  it('Renders different person icons', () => {
    mockPerson.person.kjoenn.kjoenn.value = 'M'
    wrapper.setProps({ person: mockPerson.person })
    expect(wrapper.find('Icons').props().kind).toEqual('nav-man-icon')

    mockPerson.person.kjoenn.kjoenn.value = 'X'
    wrapper.setProps({ person: mockPerson.person })
    expect(wrapper.find('Icons').props().kind).toEqual('nav-unknown-icon')

    mockPerson.person.kjoenn.kjoenn.value = 'K'
    wrapper.setProps({ person: mockPerson.person })
    expect(wrapper.find('Icons').props().kind).toEqual('nav-woman-icon')
  })
})
