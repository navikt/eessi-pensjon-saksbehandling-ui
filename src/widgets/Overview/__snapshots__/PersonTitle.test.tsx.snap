// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/Overview/PersonTitle Renders 1`] = `
<PersonTitle
  gettingPersonInfo={false}
  person={
    Object {
      "aktoer": Object {
        "ident": Object {
          "ident": "27072942618",
          "type": Object {
            "kodeRef": null,
            "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Personidenter",
            "value": "FNR",
          },
        },
      },
      "bankkonto": null,
      "bostedsadresse": Object {
        "strukturertAdresse": Object {
          "bolignummer": null,
          "gatenavn": "KJEMPEBAKKENVEIEN",
          "gatenummer": null,
          "husbokstav": null,
          "husnummer": 12,
          "kommunenummer": "1201",
          "landkode": Object {
            "kodeRef": null,
            "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Landkoder",
            "value": "NOR",
          },
          "poststed": Object {
            "kodeRef": null,
            "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Postnummer",
            "value": "5036",
          },
          "tilleggsadresse": null,
          "tilleggsadresseType": "Offisiell adresse",
        },
      },
      "diskresjonskode": null,
      "doedsdato": Object {
        "doedsdato": "2018-05-19T22:00:00.000+0000",
        "endretAv": null,
        "endringstidspunkt": null,
        "endringstype": null,
      },
      "endretAv": null,
      "endringstidspunkt": null,
      "endringstype": null,
      "foedested": null,
      "foedselsdato": Object {
        "endretAv": null,
        "endringstidspunkt": null,
        "endringstype": null,
        "foedselsdato": "1929-07-26T13:00:00.000+0000",
      },
      "geografiskTilknytning": Object {
        "geografiskTilknytning": "0213",
      },
      "gjeldendePostadressetype": Object {
        "kodeRef": null,
        "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Postadressetyper",
        "value": "POSTADRESSE",
      },
      "harFraRolleI": Array [
        Object {
          "endretAv": null,
          "endringstidspunkt": null,
          "endringstype": null,
          "harSammeBosted": false,
          "tilPerson": Object {
            "aktoer": Object {
              "ident": Object {
                "ident": "27066349360",
                "type": Object {
                  "kodeRef": null,
                  "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Personidenter",
                  "value": "FNR",
                },
              },
            },
            "bostedsadresse": null,
            "diskresjonskode": null,
            "doedsdato": null,
            "foedselsdato": null,
            "harFraRolleI": Array [],
            "kjoenn": null,
            "personnavn": Object {
              "endretAv": null,
              "endringstidspunkt": null,
              "endringstype": null,
              "etternavn": "ELQUAZZANI",
              "fornavn": "SEAFORA-MASK",
              "mellomnavn": null,
              "sammensattNavn": "ELQUAZZANI SEAFORA-MASK",
            },
            "personstatus": null,
            "postadresse": null,
            "sivilstand": null,
            "statsborgerskap": null,
          },
          "tilRolle": Object {
            "kodeRef": null,
            "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Familierelasjoner",
            "value": "BARN",
          },
        },
        Object {
          "endretAv": null,
          "endringstidspunkt": null,
          "endringstype": null,
          "harSammeBosted": false,
          "tilPerson": Object {
            "aktoer": Object {
              "ident": Object {
                "ident": "25076029647",
                "type": Object {
                  "kodeRef": null,
                  "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Personidenter",
                  "value": "FNR",
                },
              },
            },
            "bostedsadresse": null,
            "diskresjonskode": null,
            "doedsdato": null,
            "foedselsdato": null,
            "harFraRolleI": Array [],
            "kjoenn": null,
            "personnavn": Object {
              "endretAv": null,
              "endringstidspunkt": null,
              "endringstype": null,
              "etternavn": "KAEWYOO",
              "fornavn": "KAROLI-MASK",
              "mellomnavn": null,
              "sammensattNavn": "KAEWYOO KAROLI-MASK",
            },
            "personstatus": null,
            "postadresse": null,
            "sivilstand": null,
            "statsborgerskap": null,
          },
          "tilRolle": Object {
            "kodeRef": null,
            "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Familierelasjoner",
            "value": "BARN",
          },
        },
      ],
      "kjoenn": Object {
        "kjoenn": Object {
          "kodeRef": null,
          "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Kj_c3_b8nnstyper",
          "value": "K",
        },
      },
      "kontaktinformasjon": Array [],
      "maalform": null,
      "midlertidigPostadresse": null,
      "personnavn": Object {
        "endretAv": null,
        "endringstidspunkt": null,
        "endringstype": null,
        "etternavn": "HØYSÆTHER",
        "fornavn": "NAZAKMIR-MASK",
        "mellomnavn": null,
        "sammensattNavn": "HØYSÆTHER NAZAKMIR-MASK",
      },
      "personstatus": Object {
        "endretAv": null,
        "endringstidspunkt": null,
        "endringstype": null,
        "personstatus": Object {
          "kodeRef": null,
          "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Personstatuser",
          "value": "DØD",
        },
      },
      "postadresse": Object {
        "endretAv": null,
        "endringstidspunkt": null,
        "endringstype": null,
        "ustrukturertAdresse": Object {
          "adresselinje1": "NYACHAE",
          "adresselinje2": "S-BHF ORANIENBURGER STR.",
          "adresselinje3": "20095",
          "adresselinje4": null,
          "landkode": Object {
            "kodeRef": null,
            "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Landkoder",
            "value": "SWE",
          },
          "postnr": null,
          "poststed": null,
        },
      },
      "sikkerhetstiltak": null,
      "sivilstand": Object {
        "endretAv": null,
        "endringstidspunkt": null,
        "endringstype": null,
        "fomGyldighetsperiode": "2006-06-18T22:00:00.000+0000",
        "sivilstand": Object {
          "kodeRef": null,
          "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Sivilstander",
          "value": "ENKE",
        },
        "tomGyldighetsperiode": null,
      },
      "statsborgerskap": Object {
        "endretAv": null,
        "endringstidspunkt": null,
        "endringstype": null,
        "land": Object {
          "kodeRef": null,
          "kodeverksRef": "http://nav.no/kodeverk/Kodeverk/Landkoder",
          "value": "DNK",
        },
      },
      "tilrettelagtKommunikasjon": Array [],
      "vergeListe": Array [],
    }
  }
>
  <styled.div>
    <div
      className="sc-AxhUy lbrUut"
    >
      <img
        alt="nav-woman-icon"
        className="dead"
        src="icon-kvinne.png"
        width={40}
      />
      <styled.div>
        <div
          className="sc-AxirZ jpZlcz"
        />
      </styled.div>
      <Systemtittel>
        <TypografiBase
          type="systemtittel"
        >
          <h2
            className="typo-systemtittel"
          >
            HØYSÆTHER NAZAKMIR-MASK
             (
            89
            ) - 
            27072942618
          </h2>
        </TypografiBase>
      </Systemtittel>
    </div>
  </styled.div>
</PersonTitle>
`;
